## Question2: KNN Classification in sklearn

Submited by: Nikhil Dhiman 403557424

Write and submit your python codes in “Jupyter Notebook” to perform the following tasks (submit the .ipynb file).  Make sure to provide proper descriptions as MarkDown for each section of your code (each section of the code must have a short meaningful description right before that, describing what this part of the code is supposed to do!)


A. 
Read the iris flower dataset from the following URL: https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csvLinks to an external site. and assign it to a Pandas DataFrame as you learned in tutorial Lab2-3.



In [113]:
# Import Required Libraries**
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [114]:
# Will read the iris dataset from a URL into a Pandas DataFrame and explore the first few rows to understand the structure of the data.
url = "https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv"
df = pd.read_csv(url)

# Display the first few rows of the dataset to understand its structure
df.head()


sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

B. Split the dataset into testing and training sets with the following parameters: test_size=0.4, random_state=10


In [116]:
# Define features and labels
X = df.drop("species", axis=1)  # All features except 'species'
y = df["species"]               # The target variable (species)

# Split the dataset (60% training, 40% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=10)


C. Instantiate a KNN object with K=3, train it on the training set and test it on the testing set. Then, calculate the accuracy of your prediction as you learned in Lab3.

In [118]:
# Instantiate the KNN classifier with K=3
knn = KNeighborsClassifier(n_neighbors=3)

# Train the model on the training data
knn.fit(X_train, y_train)

# Make predictions on the test set
y_pred = knn.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy with K=3: {accuracy}")


Accuracy with K=3: 0.9333333333333333


D. Repeat part (c) for K=1, K=3, K=5, K=11, K=15, K=27, K=43 (you can simply use a “for loop,” and save the final accuracy results in a list). Does the accuracy always get better by increasing the value K?

In [120]:
# Define a list of K values to evaluate
k_values = [1, 3, 5, 11, 15, 27, 43]
accuracies = []

# Loop through each K, train the model, and calculate accuracy
for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    y_pred = knn.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    accuracies.append(accuracy)
    print(f"K={k}, Accuracy={accuracy}")


K=1, Accuracy=0.9166666666666666
K=3, Accuracy=0.9333333333333333
K=5, Accuracy=0.95
K=11, Accuracy=0.9666666666666667
K=15, Accuracy=0.95
K=27, Accuracy=0.95
K=43, Accuracy=0.9166666666666666


#### Does the accuracy always get better by increasing the value K?

No, increasing the value of K doesn't always improve accuracy.

In the results:

Accuracy improves when K goes from 1 to 11.
But when K increases to 15, 27, and 43, accuracy either stays the same or drops.

This happens because when K gets too large, the model starts considering too many neighbors, which can blur the boundaries between different classes, making predictions less accurate. The best K depends on the dataset, so it's important to test different values to find the one that works best.

E. Now, suppose that we would like to make prediction based on only ONE single feature. To find the best single feature, we have to try every feature individually. In other word, we have to repeat part (c) with K=11, four times (each time using only one of the 4 features), and compute the accuracy each time. Then, check the accuracies. 
Which individual feature provide the best accuracy (the best feature)?  Which one is the second best feature? (Note: you have to train, test, and evaluate your model 4 times, each time on a dataset including only one of the features, and save the final accuracy results in a list).


In [123]:
# List to store accuracy for each individual feature
individual_feature_accuracies = []

# Loop through each feature (one at a time)
for i in range(X.shape[1]):
    X_single_feature = X.iloc[:, i].values.reshape(-1, 1)
    X_train_single, X_test_single, y_train_single, y_test_single = train_test_split(X_single_feature, y, test_size=0.4, random_state=10)
    
    knn = KNeighborsClassifier(n_neighbors=11)
    knn.fit(X_train_single, y_train_single)
    y_pred_single = knn.predict(X_test_single)

    
    accuracy = accuracy_score(y_test_single, y_pred_single)
    individual_feature_accuracies.append(accuracy)
    print(f"Feature {i+1}, {X.columns[i]}, Accuracy: {accuracy}")

# Display the results for individual features
individual_feature_accuracies


Feature 1, sepal_length, Accuracy: 0.6166666666666667
Feature 2, sepal_width, Accuracy: 0.5166666666666667
Feature 3, petal_length, Accuracy: 0.95
Feature 4, petal_width, Accuracy: 0.95


[0.6166666666666667, 0.5166666666666667, 0.95, 0.95]

#### Which individual feature provide the best accuracy (the best feature)?  Which one is the second best feature?

The best accuracy is 0.95 of petal_length. 
The second best accuracy is 0.95  of petal_width.

Then third best feature is sepal_length with accuracy of 0.6166666666666667.

In [125]:
# Get the index of the best individual feature
best_single_feature_idx = individual_feature_accuracies.index(max(individual_feature_accuracies))

# Make a copy of the accuracies list and remove the best feature
individual_feature_accuracies_copy = individual_feature_accuracies.copy()
individual_feature_accuracies_copy[best_single_feature_idx] = -1  # Temporarily invalidate the best feature

# Get the second best feature by finding the next maximum
second_best_single_feature_idx = individual_feature_accuracies_copy.index(max(individual_feature_accuracies_copy))

# Get the feature names
best_feature_name = X.columns[best_single_feature_idx]
second_best_feature_name = X.columns[second_best_single_feature_idx]

# Print both the best and second best features
print(f"Best individual feature is '{best_feature_name}' with accuracy: {individual_feature_accuracies[best_single_feature_idx]}")
print(f"Second best individual feature is '{second_best_feature_name}' with accuracy: {individual_feature_accuracies[second_best_single_feature_idx]}")

Best individual feature is 'petal_length' with accuracy: 0.95
Second best individual feature is 'petal_width' with accuracy: 0.95


D. Now, we want to repeat part (e) (with K=11), this time using two features. you have to train, test, and evaluate your model for 6 different cases: using (1st and 2nd features), (1st and 3rd features), (1st and 4th  features), (2nd  and 3rd  features), (2nd and 4th features), (3rd and 4th  features)!    
Which “feature pair” provides the best accuracy?

In [127]:
# List to store accuracy for each feature pair
feature_pairs_accuracies = []

# Define feature pairs (manual selection of 6 pairs)
feature_pairs = [(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]

# Loop through each pair of features
for pair in feature_pairs:
    feature_1_name = X.columns[pair[0]]
    feature_2_name = X.columns[pair[1]]
    
    X_pair_features = X.iloc[:, [pair[0], pair[1]]]
    X_train_pair, X_test_pair, y_train_pair, y_test_pair = train_test_split(X_pair_features, y, test_size=0.4, random_state=10)
    
    knn = KNeighborsClassifier(n_neighbors=11)
    knn.fit(X_train_pair, y_train_pair)
    y_pred_pair = knn.predict(X_test_pair)
    
    accuracy = accuracy_score(y_test_pair, y_pred_pair)
    feature_pairs_accuracies.append(accuracy)
    
    print(f"Features {pair}: '{feature_1_name}' & '{feature_2_name}', Accuracy: {accuracy}")

# Display the results for feature pairs
feature_pairs_accuracies


Features (0, 1): 'sepal_length' & 'sepal_width', Accuracy: 0.7166666666666667
Features (0, 2): 'sepal_length' & 'petal_length', Accuracy: 0.9166666666666666
Features (0, 3): 'sepal_length' & 'petal_width', Accuracy: 0.9166666666666666
Features (1, 2): 'sepal_width' & 'petal_length', Accuracy: 0.95
Features (1, 3): 'sepal_width' & 'petal_width', Accuracy: 0.9666666666666667
Features (2, 3): 'petal_length' & 'petal_width', Accuracy: 0.95


[0.7166666666666667,
 0.9166666666666666,
 0.9166666666666666,
 0.95,
 0.9666666666666667,
 0.95]

In [128]:
# Find the best feature pair
best_feature_pair_idx = feature_pairs_accuracies.index(max(feature_pairs_accuracies))

# Get the names of the best feature pair
best_feature_pair_names = (X.columns[feature_pairs[best_feature_pair_idx][0]], X.columns[feature_pairs[best_feature_pair_idx][1]])

# Print the results with feature names
print(f"Best feature pair is '{best_feature_pair_names[0]}' & '{best_feature_pair_names[1]}' with accuracy: {feature_pairs_accuracies[best_feature_pair_idx]}")


Best feature pair is 'sepal_width' & 'petal_width' with accuracy: 0.9666666666666667


### Which “feature pair” provides the best accuracy?

Best feature pair is 'sepal_width' & 'petal_width' with accuracy: 0.9666666666666667


In [130]:
# Find the best and second best individual feature accuracies
best_single_feature_idx = individual_feature_accuracies.index(max(individual_feature_accuracies))

# Make a copy and remove the best feature to find the second-best
individual_feature_accuracies_copy = individual_feature_accuracies.copy()
individual_feature_accuracies_copy[best_single_feature_idx] = -1  # Temporarily invalidate the best feature
second_best_single_feature_idx = individual_feature_accuracies_copy.index(max(individual_feature_accuracies_copy))

# Find the best feature pair
best_feature_pair_idx = feature_pairs_accuracies.index(max(feature_pairs_accuracies))

# Get the names of the best and second-best individual features
best_feature_name = X.columns[best_single_feature_idx]
second_best_feature_name = X.columns[second_best_single_feature_idx]

# Get the names of the best feature pair
best_feature_pair_names = (X.columns[feature_pairs[best_feature_pair_idx][0]], X.columns[feature_pairs[best_feature_pair_idx][1]])

# Print the results with feature names
print(f"Best individual feature is '{best_feature_name}' with accuracy: {individual_feature_accuracies[best_single_feature_idx]}")
print(f"Second best individual feature is '{second_best_feature_name}' with accuracy: {individual_feature_accuracies[second_best_single_feature_idx]}")
print(f"Best feature pair is '{best_feature_pair_names[0]}' & '{best_feature_pair_names[1]}' with accuracy: {feature_pairs_accuracies[best_feature_pair_idx]}")


Best individual feature is 'petal_length' with accuracy: 0.95
Second best individual feature is 'petal_width' with accuracy: 0.95
Best feature pair is 'sepal_width' & 'petal_width' with accuracy: 0.9666666666666667


#### Question: Does the “best feature pair” from part (f) contain of both “first best feature” and “second best feature” from part (e)? In other word, can we conclude that the “best two features” for classification are the first best feature along with the second best feature together? 

Justify you  answer. If yes, why?  If no, why not



No, the "best feature pair" does not consist of both the "best individual feature" and the "second best individual feature" from part (e).

In this case:

The best individual features are 'petal_length' and 'petal_width' (both with 0.95 accuracy).
However, the best feature pair is 'sepal_width' and 'petal_width', with a higher accuracy of 0.9667.
This shows that even though 'petal_width' is part of both the best individual and pair classifications, 'petal_length' (the other best individual feature) does not appear in the best pair. Instead, 'sepal_width' appears with 'petal_width in the best feature pair.

This indicates that the combination of 'sepal_width' and 'petal_width' works better together than 'petal_length' and 'petal_width' for classification. Therefore, we cannot conclude that the "best two features" for classification are simply the top two individual features used together. Sometimes, a different combination of features provides better classification performance.?